### PR TITLE
Sk ja create edit programs

### PR DESCRIPTION
**Pull request by ja0911**
### ℹ️ Issue

Closes #[59](https://github.com/Code-4-Community/outstanding-life/issues/59)

### 📝 Description

Implemented UI for Admins to create and edit events in the Programs page

### ✔️ Verification

Tested functionality and UI on local server

Screenshot (13)

Screenshot (4)

### 🏕️ (Optional) Future Work / Notes

- Form values are yet to be stored in variables

### CODE REVIEW

File: lib/components/edit-program-modal/index.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import ProgramForm from '../program-form';
```
Comment (thejamielin) [inline lib/components/edit-program-modal/index.tsx:1]: The edit-program-modal is exactly the same format as the create-program-modal. Is there a way we can abstract this?

**Review by thejamielin — CHANGES_REQUESTED** [review]
Great progress so far! will make more reviews tomorrow

**Review by adriannaliu — COMMENTED** [review]
looking great so far guys! jamie made some good points, but other than that lgtm!

File: lib/components/header/index.tsx [hunk 2]
```diff
@@ -1,13 +1,37 @@
-import { Heading, Box } from '@chakra-ui/react';
+import { Heading, Box, Button, Text, Flex } from '@chakra-ui/react';
 
 interface HeaderProps {
   header: String;
   style?: React.CSSProperties;
+  hasButton?: Boolean;
```
[the remark this answers is not in the document]
Comment (saahilkumar) [reply lib/components/header/index.tsx]: Switched to `boolean`

File: lib/components/edit-program-modal/index.tsx (hunk 1 above)
Comment (saahilkumar) [reply lib/components/edit-program-modal/index.tsx:1]: Imo it'd be better to keep it like this cuz the `EditProgramModal` is gonna eventually autofill its values and the `CreateProgramModal` isn't, and they're also gonna call different routes when the `Submit` button is clicked. If people think there's more abstracting that can be done though, I'm down to do it.

**Review by adriannaliu — CHANGES_REQUESTED** [review]
thanks for addressing all the changes! mind taking a look at the failing check?

**Comment by yanDavid21** [comment]
nit: non-functional changes should be kept separate from functional changes (ie. .github and prettier)

File: lib/components/header/index.tsx [hunk 3]
```diff
@@ -1,15 +1,50 @@
...
+  onButtonClick,
+}: HeaderProps) => {
+  const breakpoint = useBreakpoint();
+
+  return (
+    <Box style={style}>
+      <Flex flexDir={'row'} justifyContent="space-between" mb={'3px'}>
+        <Heading>{header}</Heading>
+        {hasButton ? (
+          <Button
```
Comment (yanDavid21) [inline lib/components/header/index.tsx:26]: Perhaps a separate functional component would be helpful here so you can name this button, because right now I'd have to dig where this is called to figure out the context/meaning of the button

File: lib/components/pop-up/index.tsx [hunk 4]
```diff
@@ -15,6 +15,7 @@ interface PopUpProps {
   confirmText?: string;
   onConfirm: () => void;
   style?: React.CSSProperties;
+  size?: string;
```
Comment (yanDavid21) [inline lib/components/pop-up/index.tsx:18]: nit: you can have an enum or string union type here instead of a string ie: size?: "lg" | "sm" | "md"

File: lib/components/programs-list/index.tsx [hunk 5]
```diff
@@ -19,42 +29,69 @@ const ProgramListing: React.FC<ProgramListingProps> = ({
...
-        </Text>
-        <Link
-          href={recordingLink || registerLink}
-          isExternal
-          _hover={{ textDecoration: 'none' }}
-          _focus={{ outline: 'none' }}>
+    <>
+      <EditProgramModal modalManagement={editProgramModalManagement} />
+
+      <Flex flexDirection={useMobileLayout ? 'column' : 'row'} gap={4}>
```
Comment (yanDavid21) [inline lib/components/programs-list/index.tsx:39]: Consider breaking this down into smaller components with descriptive names?

**Review by yanDavid21 — COMMENTED** [review]
Looks good! submitted a couple bits of feedback, lmk if you agree

File: lib/components/create-program-modal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import ProgramForm from '../program-form';
+import PopUp from '../pop-up';
+
+type CreateProgramModalProps = {
+  modalManagement: {
```
[the remark this answers is not in the document]
Comment (saahilkumar) [reply lib/components/create-program-modal/index.tsx:5]: Not sure what ur asking. `isOpen`, `onOpen`, and `onClose` are just the fields of `modalManagement`. That's how ChakraUI's modals work.

File: lib/components/edit-program-modal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import PopUp from '../pop-up';
+
+type EditProgramModalProps = {
+  modalManagement: {
```
[the remark this answers is not in the document]
Comment (saahilkumar) [reply lib/components/edit-program-modal/index.tsx:5]: same as above

File: lib/components/programs-list/index.tsx (hunk 5 above)
Comment (saahilkumar) [reply lib/components/programs-list/index.tsx:39]: This is old code that just got moved around slightly, refactoring it is probs out of scope of the ticket

**Comment by saahilkumar** [comment]
>Will EditProgramModal and CreateProgramModal both have Pop and ProgramModal components? If so, I think abstraction would still be helpful if you anticipate a change to one would mean a change to another

I think they're as abstracted as they should be rn, the only similarity between the two is that they use the same form which we abstracted into `ProgramForm`. Wdym by "a change to one would mean a change to another"?

File: lib/components/create-program-modal/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (yanDavid21) [reply lib/components/create-program-modal/index.tsx:5]: I meant why not:
type CreateProgramModalProps = {
    isOpen: boolean;
    onOpen: () => void;
    onClose: () => void;
}

File: lib/components/programs-list/index.tsx (hunk 5 above)
Comment (yanDavid21) [reply lib/components/programs-list/index.tsx:39]: gotcha, agreed

File: lib/components/create-program-modal/index.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (saahilkumar) [reply lib/components/create-program-modal/index.tsx:5]: Cuz I feel like that's not very scalable. I don't anticipate any other modal props but it could still end up getting some in the future, when we implement the actual create functionality and stuff. Might as well leave that option open

File: .github/ISSUE_TEMPLATE/ticket-story.md [hunk 8]
```diff
@@ -5,11 +5,11 @@ about: A basic template for writing tickets in the form of user stories to hand
 title: ''
 labels: ''
 assignees: ''
-
 ---
 
 ### Description
-<!-- 
+
+<!--
```
Comment (thejamielin) [inline .github/ISSUE_TEMPLATE/ticket-story.md:12]: It is fine to include these minor changes formatting here, but bc it doesn't really relate to the scope of this ticket, can we exclude it in the future to minimize amount of code reviews are looking at? if want to update formatting of things, feel free to make a new ticket

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -12,9 +12,9 @@ Briefly list the changes made to the code:
 
 ### ✔️ Verification
 
-What steps did you take to verify your changes work? These should be clear enough for someone to be able to clone the branch and follow the steps themselves. 
+What steps did you take to verify your changes work? These should be clear enough for someone to be able to clone the branch and follow the steps themselves.
```
Comment (thejamielin) [inline .github/pull_request_template.md:15]: Same comment as above, let's not have minimal changes here that don't relate to scope of the ticket

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -47,7 +47,7 @@ jobs:
         with:
           # See supported Node.js release schedule at https://nodejs.org/en/about/releases/
           # We are using Node.js 14 because it is the current LTS version
-          node-version: 14.x  
+          node-version: 14.x
```
Comment (thejamielin) [inline .github/workflows/ci.yml:50]: same comment as above, can we just take out these changes

File: .prettierrc.json [hunk 11]
```diff
@@ -5,4 +5,4 @@
   "bracketSpacing": true,
   "semi": true,
```
Comment (thejamielin) [inline .prettierrc.json:6]: same comment as above

File: cypress/integration/TC01_about_us.spec.js [hunk 12]
```diff
@@ -1,17 +1,17 @@
...
-    })
-    it('should scroll to display correct content when the learn more button is clicked', () => {
-        cy.get('[data-cy=learn-more-button]').should('be.visible').click();
-        cy.get('[data-cy=mission-statement-section]').should('be.visible');
-        cy.get('[data-cy=need-section]').scrollIntoView().should('be.visible');
-        cy.get('[data-cy=leadership-team-section]').scrollIntoView().should('be.visible');
-    })
-})
\ No newline at end of file
+  beforeEach(() => {
```
Comment (thejamielin) [inline cypress/integration/TC01_about_us.spec.js:5]: how are we formatting our indentation? Was this automatically added using prettier? maybe we can get everyone to use the same formatter for our code

File: global.d.ts [hunk 13]
```diff
@@ -1,5 +1,5 @@
-import { PrismaClient } from "@prisma/client";
+import { PrismaClient } from '@prisma/client';
 
 declare global {
-    var prisma: PrismaClient;
-}
\ No newline at end of file
+  var prisma: PrismaClient;
```
Comment (thejamielin) [inline global.d.ts:4]: okay I'm noticing a lot of formatting issues here. we can keep these in, but yeah let's just make another ticket in the future to make formatting cleaner

File: lib/components/program-form/index.tsx [hunk 14]
```diff
@@ -0,0 +1,46 @@
+import { FormControl, FormLabel, Input, Textarea, Flex, Checkbox } from '@chakra-ui/react';
+
+type ProgramFormProps = {};
```
Comment (thejamielin) [inline lib/components/program-form/index.tsx:3]: Do we need to include this if it doesn't take in any props?

**Review by thejamielin — COMMENTED** [review]
still reading through this! lg so far

File: global.d.ts (hunk 13 above)
Comment (ja0911) [reply global.d.ts:4]: hi! yes so I saw David's comment from earlier too, but these were formatting fixes that needed to happen for us to pass the CI/CD checks

File: lib/components/program-form/index.tsx (hunk 14 above)
Comment (ja0911) [reply lib/components/program-form/index.tsx]: good catch!

File: lib/components/programs-list/index.tsx [hunk 15]
```diff
@@ -19,42 +29,69 @@ const ProgramListing: React.FC<ProgramListingProps> = ({
...
+            isExternal
+            _hover={{ textDecoration: 'none' }}
+            _focus={{ outline: 'none' }}>
+            <Button
+              mt="5px"
+              bg="var(--magenta)"
+              borderRadius="md"
+              alt="External page link to register for this event."
+              pointerEvents="none">
+              <Text margin="2px" color="white" fontSize="large" fontWeight="bolder">
```
Comment (lucass0717) [inline lib/components/programs-list/index.tsx:87]: it was mentioned that refactoring this is out of the scope of the ticket, but for the future is there a way to abstract the styling outside of the code for this component?

File: lib/components/edit-program-modal/index.tsx [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  };
+};
+
+const EditProgramModal: React.FC<EditProgramModalProps> = ({ modalManagement }) => {
+  return (
+    <PopUp
+      header="Edit Event"
+      manageModal={modalManagement}
+      confirmText="Edit"
+      size="2xl"
```
Comment (lucass0717) [inline lib/components/edit-program-modal/index.tsx:18]: Will the size of this remain constant regardless of window size / screen resolution?

File: lib/components/programs-list/index.tsx (hunk 15 above)
Comment (yanDavid21) [reply lib/components/programs-list/index.tsx:87]: Yeah so only commit changes to the files you've changed. If you let the linter run and do git add . it will comingle your changes with other files.
tl;dr: don't do git add . or git commit -a unless you've checked all changes are yours

File: global.d.ts (hunk 13 above)
Comment (yanDavid21) [reply global.d.ts:4]: Mentioned below, but in the future I would just omit these files from git for this PR

File: lib/components/edit-program-modal/index.tsx (hunk 16 above)
Comment (ja0911) [reply lib/components/edit-program-modal/index.tsx:18]: yes, the modal is still responsive, the '2xl' just means it'll take up the majority of the screen even when resized

File: lib/components/header/index.tsx (hunk 3 above)
Comment (thejamielin) [reply lib/components/header/index.tsx:26]: +1 to this!

File: lib/components/pop-up/index.tsx [hunk 17]
```diff
@@ -15,6 +15,7 @@ interface PopUpProps {
   confirmText?: string;
   onConfirm: () => void;
   style?: React.CSSProperties;
+  size?: 'lg' | 'sm' | 'md' | 'xl' | '2xl';
```
Comment (thejamielin) [inline lib/components/pop-up/index.tsx:18]: Do we need to specify all of these sizes yet? Aren't we just using 2xl so far? It may be a good idea to limit code to only what we need to use

File: lib/components/pop-up/index.tsx [hunk 18]
```diff
@@ -25,10 +26,15 @@ const PopUp: React.FC<PopUpProps> = ({
...
+  size,
 }) => {
   return (
     <>
-      <Modal styleConfig={style} isOpen={manageModal.isOpen} onClose={manageModal.onClose}>
+      <Modal
+        styleConfig={style}
+        isOpen={manageModal.isOpen}
+        onClose={manageModal.onClose}
+        size={size ?? 'lg'}>
```
Comment (thejamielin) [inline lib/components/pop-up/index.tsx:37]: Why did we choose to have our default to be large?

File: pages/programs/index.tsx [hunk 19]
```diff
@@ -104,6 +107,8 @@ const pastPrograms: ProgramListingProps[] = [
 ];
 
 const Programs: React.FC = () => {
+  const createProgramModalManagement = useDisclosure();
```
Comment (thejamielin) [inline pages/programs/index.tsx:110]: Did we forget to include the the CreateProgramModal on this page? I don't see it locally

image

**Review by thejamielin — CHANGES_REQUESTED** [review]
found some last couple things! i think we can get this merged as soon as we resolve these comments

File: lib/components/pop-up/index.tsx (hunk 17 above)
Comment (saahilkumar) [reply lib/components/pop-up/index.tsx:18]: I feel like it couldn't hurt to have all these values since stuff may change in the future. Plus, all the values listed in the enum are supported by chakra ui's modal component so nothing gonna break if we ever switch from 2xl to another value.

File: lib/components/pop-up/index.tsx (hunk 18 above)
Comment (saahilkumar) [reply lib/components/pop-up/index.tsx:37]: I believe it's because the modal on the donation page was size 'lg' and we didn't want to have to change it so we made that the default value. Plus, I'm pretty sure it's chakra ui's default value for modals.

File: pages/programs/index.tsx (hunk 19 above)
Comment (saahilkumar) [reply pages/programs/index.tsx:110]: Fixed that up

File: pages/programs/index.tsx [hunk 20]
```diff
@@ -104,22 +107,33 @@ const pastPrograms: ProgramListingProps[] = [
...
-          <OSLHeader
-            header="Past Online Events"
-            style={{ marginTop: '60px', marginBottom: '30px' }}
-          />
-        </div>
-        <ProgramsList programListings={pastPrograms} />
-      </Flex>
-    </Container>
+    <>
+      <CreateProgramModal modalManagement={createProgramModalManagement} />
```
Comment (thejamielin) [inline pages/programs/index.tsx:114]: Did we want to adjust the header so our implementation matches the Figma file?

Figma:
image

Current Implementation:
image

File: lib/components/programs-list/index.tsx [hunk 21]
```diff
@@ -19,42 +29,68 @@ const ProgramListing: React.FC<ProgramListingProps> = ({
...
-        <Text fontSize="lg" whiteSpace="pre-line">
-          {description}
-        </Text>
-        <Link
-          href={recordingLink || registerLink}
-          isExternal
-          _hover={{ textDecoration: 'none' }}
-          _focus={{ outline: 'none' }}>
+    <>
+      <EditProgramModal modalManagement={editProgramModalManagement} />
```
Comment (thejamielin) [inline lib/components/programs-list/index.tsx:37]: Small thing but I want to know your thoughts: In the Figma designs, we don't have a version with both buttons yet since the Admin version technically can't see the button

<details>
<summary>Figma Comparison</summary>
<br>
image
</details>

Seems like it would make sense to keep the Register here button on the Admin version
Should we distance these buttons or add an "edit" icon to make them look more distinct?

**Review by thejamielin — APPROVED** [review]
Omg fantastic work @saahilkumar and @ja0911! Please respond to the comments and you can decide whether or not you want to merge what we have now, or update the formatting

**Review by adriannaliu — APPROVED** [review]
approving this so you guys have the option of merging, but check out this [slack thread](https://c4cneuteams.slack.com/archives/C02N2ULMVNJ/p1667580621218839) regarding using icons

File: pages/programs/index.tsx (hunk 20 above)
Comment (saahilkumar) [reply pages/programs/index.tsx:175]: I'm ok with changing it, don't really have a preference. It's just like that cuz the header was recently changed to not have the line go all the way to the end

**Comment by saahilkumar** [comment]
I can add the icons in tomorrow probably

File: pages/programs/index.tsx (hunk 20 above)
Comment (adriannaliu) [reply pages/programs/index.tsx:175]: i think Al fixed the october events so that the line extends all the way to the end bc of Jamie's PR comment about that, so can we stay consistent :3

**Review by adriannaliu — APPROVED** [review]
amazing guys

**Review by thejamielin — APPROVED** [review]
this is literally fire! great job @ja0911 and @saahilkumar
